src/Chapter3.hs

```diff
@@ -479,6 +488,35 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
+data House = H1 | H2 | H3 | H4
+
+data Entertainment = Church | Library
+
+newtype Castle = Castle String
+
+type Wall = ()
+
+data City = City 
+  (Maybe (Maybe Wall, Castle))
+  Entertainment
+  [House]
+
+population :: House -> Int
+population H1 = 1
+population H2 = 2
+population H3 = 3
+population H4 = 4
+
+buildCastle :: String -> City -> City
+buildCastle nn (City c e h) = City (fmap (const $ Castle nn) <$> c) e h
+
+buildHouse :: House -> City -> City
+buildHouse h (City c e hs) = City c e (h:hs)
+
+buildWalls :: City -> Maybe City
+buildWalls (City (Just (Nothing, c)) e h) 
+  | (sum $ map population h) > 9 = Just $ City (Just (Just (), c)) e h 
+buildWalls _                     = Nothing
```

Comment:
So, it is not possible to build additional walls if the castle already has walls?